Autopilot/PathManager/Src/waypointManager.cpp

```diff
@@ -0,0 +1,357 @@
+/**
+ * Waypoint Manager Methods and Helpers Implementation
+ * Author: Dhruv Rawat
+ * Created: November 2020
+ * Last Updated: November 2020 (Dhruv)
+ */
+
+#include "waypointManager.hpp"
+
+
+/*** INITIALIZATION ***/
+
+
+WaypointManager::WaypointManager(_PathData *initialWaypoints, int numberOfWaypoints, _PathData *currentLocation) {
+    
+    homeBase = _PathData(); //Initializes homebase first so there are no undefined behaviour issues
+    homeBase = *currentLocation;
+
+    numWaypoints = numberOfWaypoints;
+    nextFilledIndex = 0;
+    nextAssignedId = 0;
+    
+    // Initializes the waypointBuffer array
+    for (int i = 0; i < numWaypoints; i++) {
+        // Links waypoints together
+        if (i == 0) { // If first waypoint, link to next one only
+            initialWaypoints[i].next = &initialWaypoints[i+1]; // Passes address so any updates in initialWaypoints[i+1] is reflected in initialWaypoints[i].next
+        } else if (i == numWaypoints - 1) { // If last waypoint, link to previous one only
+            initialWaypoints[i].previous = &initialWaypoints[i-1];
+            nextAssignedId = initialWaypoints[i].waypointId + 1; // Gets last assigned ID and increments by 1 :))
+        } else {
+            initialWaypoints[i].next = &initialWaypoints[i+1];
+            initialWaypoints[i].previous = &initialWaypoints[i-1];
+        }
+        
+        waypointBuffer[i] = &initialWaypoints[i]; // Sets the element in the waypointBuffer
+        waypointBufferStatus[i] = FULL;
+        nextFilledIndex = i + 1;
+    }
+
+    errorStatus = WAYPOINT_SUCCESS;
+}
+
+_PathData* WaypointManager::initialize_waypoint() {
+    _PathData* waypoint = new _PathData(); // Create new waypoint in the heap
+    waypoint->waypointId = nextAssignedId++; // Set ID and increment
+    // Set next and previous waypoints to empty for now
+    waypoint->next = 0; 
+    waypoint->previous = 0;
+    return waypoint;
+}
+
+_PathData* WaypointManager::initialize_waypoint(long double longitude, long double latitude, int altitude, int waypointType) {
+    _PathData* waypoint = new _PathData(); // Create new waypoint in the heap
+    waypoint->waypointId = nextAssignedId++; // Set ID and increment
+    waypoint->latitude = latitude;
+    waypoint->longitude = longitude;
+    waypoint->altitude = altitude;
+    waypoint->waypointType = waypointType;
+    // Set next and previous waypoints to empty for now
+    waypoint->next = 0; 
+    waypoint->previous = 0;
+    return waypoint;
+}
+
+_PathData* WaypointManager::initialize_waypoint(long double longitude, long double latitude, int altitude, int waypointType, float turnRadius) {
```

Comment:
These are great! Only trouble I see is that to use them, you have to have already created an object, which would have required the user to input initial way points in the constructor. I'd consider making these static methods. Makes a lot of sense to me, since they're not really part of of the path management engine as much as a helpful tool.

Reply:
Yeah, I realized that as I was debugging. I will def transition to this soon!

Reply:
My facepalm when I noticed it sent my brain to mars.